web3/method.py

```diff
@@ -0,0 +1,144 @@
+from eth_utils import (
+    to_tuple,
+)
+from eth_utils.toolz import (
+    identity,
+    pipe,
+)
+
+from web3.exceptions import (
+    UndefinedMethodError,
+)
+
+TEST_METHOD_CONFIG = {
+    'name': "signInBlood",
+    'mungers': [],
+    'json_rpc_method': "eth_signTransactionInBlood",
+}
+
+
+def lookup_method(module, module_config, method_class, attr_name):
+    try:
+        method = module_config[attr_name]
+    except KeyError:
+        raise UndefinedMethodError("No method named {0}".format(attr_name))
+    return method_class(module.web3, method)
+
+
+class DummyRequestManager:
+    def request_blocking(method, params):
+        return (method, params)
+
+
+class DummyWeb3:
+    manager = DummyRequestManager
+
+
+def test_sync_method_config_loading():
+    signInBlood = BlockingMethod(DummyWeb3(), TEST_METHOD_CONFIG)
+    signInBlood.input_munger = lambda *_: {}
+    assert signInBlood.method_selector
+    assert ('eth_signTransactionInBlood', {}) == signInBlood()
+
+
+class BaseMethod:
+    """BaseMethod for web3 module methods
+
+    Calls to the Method go through these steps:
+
+    1. input munging ;) - includes normalization, parameter checking, formatters.
+    Any processing on the input parameters that need to happen before json_rpc
+    method string selection occurs.
+
+    2. method selection - function that selects the correct rpc_method. accepts a
+    function or an string.
+
+    3. constructing formatter middlewares - takes the rpc_method and looks up the
+    corresponding input/output formatters. these are the middlewares migrated here.
+
+    4. making the request through the middleware (pipeline)? wrapped request
+    function.
+    """
+    def __init__(self, web3, method_config):
+        self.__name__ = method_config.get('name', 'anonymous')
+        self.__doc__ = method_config.get('doc', '')
+        self.is_property = method_config.get('is_property', False)
```

Comment:
What do you think about removing this and only implementing the concept of direct `property` lookups at the `Module` level?

```python
class VersionModule:
    get_version = VersionMethod(...)

    @property
    def version(self):
        return self.get_version()
```

Would work similarly for `async`